src/backend/distributed/utils/shardinterval_utils.c

```diff
@@ -332,13 +342,13 @@ SearchCachedShardInterval(Datum partitionColumnValue, ShardInterval **shardInter
 
 		if (DatumGetInt32(maxValueComparison) <= 0)
 		{
-			return shardIntervalCache[middleIndex];
+			return middleIndex;
 		}
 
 		lowerBoundIndex = middleIndex + 1;
 	}
 
-	return NULL;
+	return INVALID_SHARD_INDEX;
```

Comment:
Given that this function could potentially return `INVALID_SHARD_INDEX` for `hash` distributed tables, I'm trying to figure out an example that exercise that.
However, that seems impossible given that we have shards that spans all values.(That could happend if there are bugs on our side, such as missing shards, not sorted shard cache etc.) ). 

So, should we add an `Assert (partMethod != DISTRIBUTE_BY_HASH)`  just before return and update the function comment accordingly? So that callers do not need to consider what happens in case of `INVALID_SHARD_INDEX` for hash distributed tables? Even, instead of an `Assert`, we could throw an error which is more useful than an `Assert` and can be added regression tests using manual operations on the metadata.

(For example, while creating co-located shard placements on `create_distributed_table`, the caller has nothing to do.)